### PR TITLE
refactor: search's destructor now cleans up all of its memory

### PR DESCRIPTION
**Pull request by jackgerrits**
With this PR, all reduction memory cleanup is done in destructors and therefore the all destructor can be safely used if all you desire is to destruct and don't care about model saving or reporting.

### CODE REVIEW

File: vowpalwabbit/search.h [hunk 1]
```diff
@@ -83,24 +83,24 @@ struct search
   template <class T>
   void set_task_data(T* data)
   {
-    task_data = data;
+    task_data = std::shared_ptr<T>(data);
```
Comment (peterychang) [inline vowpalwabbit/search.h:86]: can these be unique_ptr instead?

**Review by peterychang — COMMENTED** [review]
really the biggest question I have is, can those shared_ptrs be converted to unique_ptrs? Or do we containing objects around?

File: library/libsearch.h [hunk 2]
```diff
@@ -27,9 +30,7 @@ template<class INPUT, class OUTPUT> class SearchTask
     d->run_f = _search_run_fn;
     d->run_setup_f = _search_setup_fn;
     d->run_takedown_f = _search_takedown_fn;
-    d->run_object = this;
-    d->extra_data  = NULL;
-    d->extra_data2 = NULL;
+    d->run_object = std::shared_ptr<SearchTask<INPUT, OUTPUT>>(this);
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply library/libsearch.h:33]: You cannot have a shared_ptr<void>

File: library/libsearch.h [hunk 3]
```diff
@@ -51,41 +52,36 @@ template<class INPUT, class OUTPUT> class SearchTask
...
   multi_ex trigger;
+  INPUT _input;
+  OUTPUT _output;
 
   void call_vw(INPUT& input_example, OUTPUT& output)
-  { HookTask::task_data* d = sch.template get_task_data<HookTask::task_data> (); // ugly calling convention :(
-    d->extra_data  = (void*)&input_example;
-    d->extra_data2 = (void*)&output;
+  {
+    _input = input_example;
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply library/libsearch.h:60]: This is in the example/library directory, I think its fine

File: vowpalwabbit/search.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/search.h:86]: No this is a shared_ptr<void>

File: library/libsearch.h [hunk 4]
```diff
@@ -51,41 +52,36 @@ template<class INPUT, class OUTPUT> class SearchTask
...
   }
 
   static void _search_run_fn(Search::search&sch)
   { HookTask::task_data* d = sch.get_task_data<HookTask::task_data>();
-    if ((d->run_object == NULL) || (d->extra_data == NULL) || (d->extra_data2 == NULL))
-    {
-      THROW("error: calling _search_run_fn without setting run object");
-    }
-    ((SearchTask*)d->run_object)->_run(sch, *(INPUT*)d->extra_data, *(OUTPUT*)d->extra_data2);
+    if (d->run_object == nullptr) { THROW("error: calling _search_run_fn without setting run object"); }
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply library/libsearch.h:67]: They are values now

**Review by lalo — APPROVED** [review]

**Comment by lalo** [comment]
Is there a python test that tests this? or I believe thats where teh requirement from shared_ptr is ocming from

**Comment by jackgerrits** [comment]
Hook task is used in Python yeah, it stores the type erases py::object. There are notebooks, but I don't think there is a test

**Comment by jackgerrits** [comment]
@peterychang, shared_ptrs are definitely needed because unique_ptrs cant type erase

**Pull request merged**